R/sgdnet.R

```diff
@@ -69,55 +91,92 @@ sgdnet.default <- function(x,
+  if (is.null(variable_names))
+    response_names <- paste0("y", seq_len(NCOL(y)))
+
   y <- as.matrix(y)
 
+  # Collect sgdnet-specific options for debugging and more
   debug <- getOption("sgdnet.debug")
 
+  if (is.null(lambda))
+    lambda <- double(0L)
+
   stopifnot(identical(NROW(y), NROW(x)),
             !any(is.na(y)),
             !any(is.na(x)),
             alpha >= 0 && alpha <= 1,
+            length(alpha) == 1L,
             thresh > 0,
-            lambda >= 0,
+            all(lambda >= 0),
             is.logical(intercept),
             is.logical(standardize),
             is.logical(debug))
 
-  # Setup reponse type options and assert that input is correct
-  switch(
-    match.arg(family),
-    gaussian = {
-      stopifnot(is.numeric(y),
-                identical(NCOL(y), 1L))
-    }
-  )
+  # Setup reponse type options and assert appropriate input
+  family <- match.arg(family)
+
+  switch(family,
+         gaussian = {
+           stopifnot(is.numeric(y),
+                     identical(NCOL(y), 1L))
+           }
+         )
+
+  control <- list(family = family,
+                  intercept = intercept,
+                  is_sparse = is_sparse,
+                  lambda = lambda,
+                  elasticnet_mix = alpha,
+                  n_lambda = nlambda,
+                  lambda_min_ratio = lambda.min.ratio,
+                  normalize = standardize,
+                  max_iter = maxit,
+                  tol = thresh,
+                  debug = debug)
 
   # Fit the model by calling the Rcpp routine.
-  res <- FitModel(x,
-                  y,
-                  family,
-                  intercept,
-                  is_sparse,
-                  alpha_sklearn,
-                  beta_sklearn,
-                  standardize,
-                  maxit,
-                  thresh,
-                  debug)
+  res <- SgdnetCpp(x, y, control)
 
-  variable_names <- colnames(x)
+  # Setup return values
+  a0 <- drop(t(as.matrix(res$a0)))
+  beta <- res$beta
+  lambda <- drop(res$lambda)
 
-  # Organize return values
-  a0 <- drop(res$a0)
-  names(a0) <- paste0("s", seq_along(a0) - 1L)
-  beta <- Matrix::Matrix(res$beta)
-  dimnames(beta) <- list(colnames(x), names(a0))
+  path_names <- paste0("s", seq_along(lambda) - 1L)
+
+  beta <- lapply(seq(dim(beta)[2L]),
+                 function(x) Matrix::Matrix(as.matrix(beta[ , x, ]),
+                                            dimnames = list(variable_names,
+                                                            path_names),
+                                            sparse = TRUE))
+
+  if (family %in% c("gaussian", "binomial", "poisson", "cox")) {
+    # NOTE(jolars): I would rather not to this, i.e. have different outputs
+    # depending on family, but this makes it equivalent to glmnet output
+    beta <- beta[[1L]]
+    names(a0) <- path_names
+  } else {
+    a0 <- t(a0)
+    rownames(a0) <- response_names
+    colnames(a0) <- path_names
+  }
```

Comment:
Okay, yeah, that was the purpose of this. I guess it would be alright if I also change the return type of the coefficients? I am personally not happy that the coefficients are sometimes a list (mgaussian and multinomial) and sometimes a matrix (as well as the direction of that matrix, having the features as rows) and would much rather always return a 3-dimensional array (cube).

This is really part of a greater discussion regarding the package and I'll post an issue so that we can discuss if and when we would like to depart from glmnet.

Reply:
I agree the `glmnet` design is weird here, but I understand what they're going for - sparse 3d array libraries are rare (`R` recently got them through the `slam` package and Armadillo doesn't have a sparse cube type). 

I lean towards sticking to `glmnet`-formatted stuff for the output of accessor functions (`coef`) while saying that we have the freedom to define internal structures however we want, but happy to discuss in a different issue.